pkg/storage/registry/spaces/spaces.go

```diff
@@ -469,20 +473,42 @@ func (r *registry) findProvidersForAbsolutePathReference(ctx context.Context, pa
 	var deepestMountPathProvider *registrypb.ProviderInfo
 	providers := map[string]*registrypb.ProviderInfo{}
 	for address, provider := range r.c.Providers {
+
+		// Only list spaces on this provider if the request matches the mountpoint
+		var providerMatchesPath bool
+		for _, sc := range provider.Spaces {
+			match, err := regexp.MatchString(sc.MountPoint, path)
+			if (err == nil && match) || strings.HasPrefix(sc.MountPoint, path) {
+				providerMatchesPath = true
+				break
+			}
+		}
+		if !providerMatchesPath {
+			continue
+		}
+
 		p := &registrypb.ProviderInfo{
 			Opaque:  &typesv1beta1.Opaque{Map: map[string]*typesv1beta1.OpaqueEntry{}},
 			Address: address,
 		}
 		var spaces []*providerpb.StorageSpace
 		var err error
-		filters := []*providerpb.ListStorageSpacesRequest_Filter{}
+
 		// when listing paths also return mountpoints
-		filters = append(filters, &providerpb.ListStorageSpacesRequest_Filter{
-			Type: providerpb.ListStorageSpacesRequest_Filter_TYPE_SPACE_TYPE,
-			Term: &providerpb.ListStorageSpacesRequest_Filter_SpaceType{
-				SpaceType: "+mountpoint",
+		filters := []*providerpb.ListStorageSpacesRequest_Filter{
+			{
+				Type: providerpb.ListStorageSpacesRequest_Filter_TYPE_PATH,
+				Term: &providerpb.ListStorageSpacesRequest_Filter_Path{
+					Path: path,
+				},
 			},
-		})
+			{
+				Type: providerpb.ListStorageSpacesRequest_Filter_TYPE_SPACE_TYPE,
+				Term: &providerpb.ListStorageSpacesRequest_Filter_SpaceType{
+					SpaceType: "+mountpoint",
+				},
+			},
+		}
 
 		spaces, err = r.findStorageSpaceOnProvider(ctx, p.Address, filters, unrestricted)
 		if err != nil {
```

Comment:
@butonic @kobergj the checks below this must be moved to the storage drivers which must have all the logic to return only the correct spaces. I can't remove this code as the tests running on decomposed fs will fail. Can we merge this PR and you can create another one for that? It'll unblock us